### ISSUE TITLE
Serialization Benchmarking

### ISSUE THREAD

**Issue by rdspring1** [NVIDIA/Fuser#679]
This issue tracks which parts of the fusion cache hierarchy to prioritize during serialization.

Comments:
- [X] Loading kernel database is non-trivial. Fixed via https://github.com/NVIDIA/Fuser/pull/814
- [ ] Most of the time spent in `GpuLower::lower` is spent in `IndexLowering::getIndexedExprs` and `ConditionalFromPredicateModifier`. Prototype PR via https://github.com/NVIDIA/Fuser/pull/558
- [ ] Scheduling the fusion and creating its Heuristics are slow.

**Comment by rdspring1** [comment]
```python
N = 1024
x = torch.randn([2048, N], device="cuda", requires_grad=True)
y = torch.randn([2048, N], device="cuda", requires_grad=True)
w = torch.ones([N], device="cuda", requires_grad=True)
b = torch.zeros([N], device="cuda", requires_grad=True)

def func(x, z, weight, bias):
    y = dropout(x)
    alpha = y + z
    return layer_norm(alpha, [N], weight, bias)
```
Trace: [add_dropout_layernorm.zip](https://github.com/NVIDIA/Fuser/files/12245456/add_dropout_layernorm.zip)

# Trace for `dropout_add_layernorm`
## Forward
![add_dropout_layernorm_fwd](https://github.com/NVIDIA/Fuser/assets/3637896/b2c18840-ecfb-4ed1-9507-2e58af31544b)


## Backward
![add_dropout_layernorm_bwd](https://github.com/NVIDIA/Fuser/assets/3637896/8dd9f5ff-e013-4aa6-b4af-d2ba78cdb2fa)

**Comment by rdspring1** [comment]
Archived trace from Aug. 2nd: [nanogpt_tracing.zip](https://github.com/NVIDIA/Fuser/files/12244255/serde_tracing.zip) --- The zip file contains the benchmark python scripts and traces.

# Trace for `test_nanogpt_mha_dpa`
## Without KernelDB
* Persistent scheduler
![trace_nanogpt_mha_dpa](https://github.com/NVIDIA/Fuser/assets/3637896/f4a4eb38-027a-49e2-a082-5fd2b7415e17)

## With KernelDB
![trace_nanogpt_mha_dpa_kernel_db](https://github.com/NVIDIA/Fuser/assets/3637896/287492b4-55b9-4e7d-96a3-2276be348e4e)

# Trace for `test_nanogpt_mha_linears`
## Without KernelDB
* Pointwise scheduler
![trace_nanogpt_mha_linears_p0](https://github.com/NVIDIA/Fuser/assets/3637896/483132ef-d08d-4a6a-90ca-dd31ebe90032)

**Comment by rdspring1** [comment]
Baseline serialization handles python frontend, fusion executor cache, and kernel database.
The `FastGpuLower` prototype is **75% faster** than baseline.

Trace Reference: [serde_benchmarking_all_python_frontend.zip](https://github.com/NVIDIA/Fuser/files/12743143/serde_benchmarking_all_python_frontend.zip)

### Baseline
![current_serialization_9_25_23](https://github.com/NVIDIA/Fuser/assets/3637896/051cb860-f40b-46c6-9b05-6e7c97c28e49)
### FastGpuLower
![fast_gpulower_pr_558](https://github.com/NVIDIA/Fuser/assets/3637896/5e093814-9f34-4ea1-a554-bbb0d0c24473)
